### PR TITLE
Add support for archived jobs

### PR DESCRIPTION
**Pull request by Martchus**
* See https://progress.opensuse.org/issues/91347 and particular commit messages
* Still WIP and related ticket is only "spike" ticket

### CODE REVIEW

File: dbicdh/PostgreSQL/upgrade/92-93/001-auto.sql
Comment (okurz) [inline dbicdh/PostgreSQL/upgrade/92-93/001-auto.sql:7]: "is_archived" sounds more like a method. I think in other cases of boolean columns we did not add the "is_" prefix.

File: lib/OpenQA/Schema/Result/Jobs.pm
Comment (okurz) [inline lib/OpenQA/Schema/Result/Jobs.pm:2082]: interesting. Why this change?

**Review by okurz — COMMENTED** [review]
looks nice. Have you found a good place where to trigger the archiving? E.g. following our idea in the ticket: Should we consider "archived" == "important" after some grace time when an important job is moved to a potentially slower archive? So at the time when we trigger the cleanup for non-important jobs we also look at the important jobs and make sure that they are archived or queued for archiving (if that's a more costly process)

File: dbicdh/PostgreSQL/upgrade/92-93/001-auto.sql
Comment (Martchus) [reply dbicdh/PostgreSQL/upgrade/92-93/001-auto.sql:7]: DBIx maps it to a method so I thought it would be a good fit. We already do the same for the boolean columns within the users table (`is_operator`, `is_admin`).

File: lib/OpenQA/Schema/Result/Jobs.pm
Comment (Martchus) [reply lib/OpenQA/Schema/Result/Jobs.pm:2082]: see https://github.com/os-autoinst/openQA/pull/3859

File: dbicdh/PostgreSQL/upgrade/92-93/001-auto.sql
Comment (Martchus) [reply dbicdh/PostgreSQL/upgrade/92-93/001-auto.sql:7]: To be fair, there are also other boolean columns *not* using that prefix. I don't really care but would postpone polishing like this until we decide that we actually want to merge the spike solution.

**Comment by Martchus** [comment]
No, triggering the archiving is not included yet. However, the idea from the ticket / our previous discussion would make sense. Of course I'd make it a configuration option (and before enabling it one needed to ensure that the archive directory/mountpoint actually exist). And I would likely make it an extra Minion job (triggered by the the cleanup Minion job) because the cleanup job is already long-running so it would make sense to split it up.

There's also some polishing missing but considering it is a spike solution I'd like to do this only at the very end (when we decide that we actually like to merge it):

* Rename the table columns to remove `is_` prefix if that's generally preferred (amending the commits I've created so far)
* Indication within the UI
* Documentation
* Test coverage of error case

**Comment by Martchus** [comment]
Adding a Minion task for this actually raised some questions, see `lib/OpenQA/Task/Job/ArchiveResults.pm`. The last commit is incomplete (only covers the config option so far).

File: dbicdh/PostgreSQL/upgrade/92-93/001-auto.sql
Comment (okurz) [reply dbicdh/PostgreSQL/upgrade/92-93/001-auto.sql:7]: sure, not important right now. Note for later :)

File: lib/OpenQA/Task/Job/ArchiveResults.pm
Comment (okurz) [inline lib/OpenQA/Task/Job/ArchiveResults.pm:28]: how about we simply ignore non-finalized job results (can we detect that?) as they should be found as archiving candidates the next time

File: lib/OpenQA/Task/Job/ArchiveResults.pm
Comment (okurz) [inline lib/OpenQA/Task/Job/ArchiveResults.pm:29]: do you mean results or screenshots cleanup? Because when triggering synchronously with results cleanup then that at least should not interfer. And screenshots are symlinks which should still be valid after archiving a job unless any screenshot cleanup happens in parallel to the directory copy, right? In that case, how about a minion lock on each corresponding job?

File: lib/OpenQA/Task/Job/ArchiveResults.pm
Comment (okurz) [inline lib/OpenQA/Task/Job/ArchiveResults.pm:31]: good point. I wonder how that would break and if we can intervene within the results tab loading to force-refresh loading that page?

File: lib/OpenQA/Task/Job/ArchiveResults.pm
Comment (Martchus) [reply lib/OpenQA/Task/Job/ArchiveResults.pm:28]: > can we detect that?

That's a good question. I think we'd have to check whether there are any `.txt` files within the results dir.

> as they should be found as archiving candidates the next time

If finalizing a job fails for some reason it is not automatically re-attempted. So we should at least try to re-attempt the finalizing.

File: lib/OpenQA/Task/Job/ArchiveResults.pm
Comment (Martchus) [reply lib/OpenQA/Task/Job/ArchiveResults.pm:29]: > do you mean results or screenshots cleanup? Because when triggering synchronously with results cleanup then that at least should not interfer.

I mean the result cleanup. I suppose it wouldn't interfere in practice but I can think of some contrived examples, e.g. a job is no longer considered important during the cleanup and then would be deleted after all (while the archiving is ongoing).

> And screenshots are symlinks which should still be valid after archiving a job unless any screenshot cleanup happens in parallel to the directory copy, right?

Yes, screenshots are not touched by this. And the symlinks remain valid as long as they use absolute paths which they luckily do. If screenshot cleanup happens in the background that shouldn't be a problem because it wouldn't delete screenshots used by existing jobs (regardless of their archiving state).

File: lib/OpenQA/Task/Job/ArchiveResults.pm
Comment (Martchus) [reply lib/OpenQA/Task/Job/ArchiveResults.pm:31]: I also wonder how that would break. Only HTTP requests which are pending when deleting the old results directory and had already been started before the job was marked as archived would cause problems.

We have controller code reading result files and some files are served directly by Apache. Likely the deletion of the original files during the archiving would fail if the result file if there are still file handles open for reading at this point. So we would likely see failed Minion jobs and leftover results. We could make it more resilient against that problem by trying to delete the results again after a short delay.

If it was actually possible to delete the files while there are still file handles open for reading I suppose the readers would run into I/O errors and at least some parts of the details tab wouldn't be shown correctly.

**Comment by Martchus** [comment]
There's an even bigger problem. So far I've only tested this using Mojolicious HTTP server directly. However, when using the Apache reverse proxy setup loading images fails because Apache would still expect them within the usual results directory.

File: lib/OpenQA/Task/Job/ArchiveResults.pm
Comment (okurz) [reply lib/OpenQA/Task/Job/ArchiveResults.pm:36]: also see https://github.com/os-autoinst/openQA/pull/3861

**Review by okurz — CHANGES_REQUESTED** [review]

File: lib/OpenQA/Task/Job/ArchiveResults.pm
Comment (Martchus) [reply lib/OpenQA/Task/Job/ArchiveResults.pm:36]: I'd prefer to do this kind of refactoring separately.

Comment (okurz) [reply lib/OpenQA/Task/Job/ArchiveResults.pm:36]: yes, this is why I created my PR. You can ignore the suggestions for here now

Comment (Martchus) [reply lib/OpenQA/Task/Job/ArchiveResults.pm:36]: If the refactoring PR gets merged after this i can still apply it here as a separate commit.

**Comment by okurz** [comment]
by the way, https://github.com/os-autoinst/openQA/pull/3862 is ready to reviewed

**Comment by Martchus** [comment]
Ticket for finalizing the PR: https://progress.opensuse.org/issues/91782

File: dbicdh/PostgreSQL/upgrade/92-93/001-auto.sql
Comment (kraih) [reply dbicdh/PostgreSQL/upgrade/92-93/001-auto.sql:7]: I'd also just call this column `archived`.

**Comment by okurz** [comment]
Looks great. Can you split out the database migration into a separate PR to merge that earlier?

**Comment by Martchus** [comment]
I've been creating https://github.com/os-autoinst/openQA/pull/3897 but I still need to check the FIXMEs.

File: lib/OpenQA/Schema/Result/Jobs.pm
Comment (okurz) [inline lib/OpenQA/Schema/Result/Jobs.pm:277]:
```suggestion
    return undef unless my $normal_result_dir = $self->archivable_result_dir;
```

if you like :) That's what we use already in other places

File: lib/OpenQA/Task/Job/ArchiveResults.pm
Comment (okurz) [inline lib/OpenQA/Task/Job/ArchiveResults.pm:34]:
```suggestion
      unless my $guard1 = $app->minion->guard('archiving_results', ONE_DAY, {limit => 5});
```

File: lib/OpenQA/Task/Job/ArchiveResults.pm
Comment (okurz) [reply lib/OpenQA/Task/Job/ArchiveResults.pm:36]: you can use `ONE_DAY` here now as well

File: lib/OpenQA/Task/Job/Limit.pm
Comment (okurz) [inline lib/OpenQA/Task/Job/Limit.pm:67]:
```suggestion
    my %options = (priority => 0, ttl => 2 * ONE_DAY);
```

**Review by okurz — CHANGES_REQUESTED** [review]

**Comment by Martchus** [comment]
See the commit message of the last commit for some caveats.

**Review by okurz — APPROVED** [review]
Approved. Please let mergify do some work, i.e. don't merge manually to give a bit more time to reviewers as well as to ensure all required checks are actually passed :)

**Review by Amrysliu — APPROVED** [review]
LGTM

**Comment by AdamWill** [comment]
should we add the archive directory to the package spec?

**Comment by Martchus** [comment]
Likely one would create a bind mount there or configure a different path because the idea is to make use of a different storage. So this would likely be not that useful considering one needs to do some manual setup anyways.

**Comment by AdamWill** [comment]
hmm, yeah, that makes sense.

**Pull request merged**
